src/findone-parallel-performance.c

```diff
@@ -0,0 +1,239 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Tests performance of parallel find operations on a client pool.
+ * The task definition mirrors the workload identified in CDRIVER-4002.
+ * The task definition is not part of the "MongoDB Driver Performance
+ * Benchmarking" specification. */
+
+#include "mongo-c-performance.h"
+
+#include <bson.h>
+#include <mongoc.h>
+
+#include <pthread.h>
+
+typedef struct {
+   pthread_t thread;
+   mongoc_client_t *client;
+   int n_findone_to_run;
+} findone_parallel_thread_context_t;
+
+typedef struct {
+   perf_test_t base;
+   mongoc_client_pool_t *pool;
+   int n_threads;
+   findone_parallel_thread_context_t *contexts;
+} findone_parallel_perf_test_t;
+
+/* FINDONE_FILTER_SIZE is the size of the BSON document {"id": 0}.
+ * Determined with this Python script:
+ *    import bson
+ *    print (len(bson.encode({"id": 0}))) # prints "13"
+ */
+#define FINDONE_FILTER_SIZE 13
+/* FINDONE_COUNT is the total number of "find" operations done by all threads.
+ * Each thread runs FINDONE_COUNT / n_threads "find" operations. */
+#define FINDONE_COUNT 10000
+
+static void *
+_findone_parallel_perf_thread (void *p)
+{
+   findone_parallel_thread_context_t *ctx =
+      (findone_parallel_thread_context_t *) p;
+   mongoc_collection_t *coll;
+   int i;
+   bson_t filter = BSON_INITIALIZER;
+
+   coll = mongoc_client_get_collection (ctx->client, "perftest", "coll");
+   bson_append_int32 (&filter, "_id", 3, 0);
+
+   for (i = 0; i < ctx->n_findone_to_run; i++) {
+      mongoc_cursor_t *cursor;
+      const bson_t *doc;
+      bson_error_t error;
+
+      cursor = mongoc_collection_find_with_opts (
+         coll, &filter, NULL /* opts */, NULL /* read_prefs */);
+      /* iterate once to send "find" */
+      if (mongoc_cursor_next (cursor, &doc)) {
+         MONGOC_ERROR (
+            "Error: unexpected document returned from collection: %s",
+            bson_as_json (doc, NULL));
+         abort ();
+      }
+      if (mongoc_cursor_error (cursor, &error)) {
+         MONGOC_ERROR ("Error from cursor iteration: %s", error.message);
+         abort ();
+      }
+      mongoc_cursor_destroy (cursor);
+   }
+
+   mongoc_collection_destroy (coll);
+   bson_destroy (&filter);
+   return NULL;
+}
+
+static void
+findone_parallel_perf_task (perf_test_t *test)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+   int i;
+   int ret;
```

Comment:
Variable "ret" is used for convenience to check and /report/ errors from pthread operations. Ok.